### ISSUE TITLE
Sidecar resource not updated properly

### ISSUE THREAD

**Issue by howardjohn** [istio/istio#11762]
**Describe the bug**
When a Sidecar resource is modified, it does not seem to update correctly.

**Expected behavior**
Applying a new config should have the same effect as deleting the config, then applying the new config, but it does not.

**Steps to reproduce the bug**
Turned on `global.outboundTrafficPolicy.mode=ALLOW_ANY` (not sure if related)

1. No Sidecar - all external traffic is blocked (`curl google.com` 404s)
2. Apply Sidecar resource
```yaml
apiVersion: networking.istio.io/v1alpha3
kind: Sidecar
metadata:
  name: default
spec:
  egress:
    - hosts:
      - ./*
```
After this external traffic works

3. `kubectl apply` same config, changed to `hosts: []` - external traffic still works
4. `kubectl remove` the Sidecar - traffic blocked
5. `kubectl apply` the config with  `hosts: []` - traffic blocked

Step 3 and 5 should have the same result.

**Version**
Istio 1.1 Snapshot 6

**Installation**
Helm

**Environment**
GKE

@rshriram

**Comment by hzxuzhonghu** [comment]
I think the title should be updated to be clearly  `Sidecar resource influence outbound traffic`

**Comment by rshriram** [comment]
step 5 implies that the code generating the config is working as expected. step 3 implies that something in the system is ignoring the config update notifications. This part of the system is in ads.go or discovery.go where there is some dampening logic.. cc @costinm who knows the details of this..
To me, 3 seems like a config update is being ignored. I have seen this issue surface every now and then with other networking configs as well.

**Comment by hzxuzhonghu** [comment]
I think we do not allow empty Hosts
```
		if len(i.Hosts) == 0 {
			errs = appendErrors(errs, fmt.Errorf("sidecar: egress listener must contain at least one host"))
		}
```

**Comment by hzxuzhonghu** [comment]
I donot know why 4 influence the rxternal traffic.

**Comment by rshriram** [comment]
./* is valid @hzxuzhonghu ..

**Comment by mbanikazemi** [comment]
Not specifying any host for egress as described in step 3 above, makes the sidecar config invalid but we do not catch it. PR #12233 adds sidecars to the resources being monitored by the validating web hook in Galley to address this.

**Comment by ozevren** [comment]
*Without* the fix in #12233, it looks like:
- Valid sidecar gets applied. (Step 2). This goes through Galley and Pilot and gets processed.
- Invalid sidecar gets applied (Step 3). Config Validator misses it, but it gets filtered out in Pilot's validation. Internal config state of Pilot doesn't change.
- Resource is removed (Step 4). Valid configuration change. Pilot changes its internal state to remove and push.
- Reapply  invalid sidecar (Step 5). Config validator misses, it, but it gets filtered out in Pilot's validation. Internal config state doesn't change (i.e. its empty).

I tried this locally with #12233 applied, and it looks like config validator is filtering out and returning an error during kubectl apply.

Am I missing anything? I don't see any other issues in this particular bug? (Feel free to correct me).

If there are no other issues? Can we close this one?

**Comment by howardjohn** [comment]
Sounds like it is fixed! Closing this now
